lighthouse-core/lib/network-recorder.js

```diff
@@ -99,6 +99,18 @@ class NetworkRecorder extends EventEmitter {
     return !!(isQUIC && receivedHeaders && record.endTime);
   }
 
+  /**
+   * frame root network requests don't always "finish" even when they're done loading data, use responseReceived instead
+   * @see https://github.com/GoogleChrome/lighthouse/issues/6067#issuecomment-423211201
+   * @param {LH.Artifacts.NetworkRequest} record
+   * @return {boolean}
+   */
+  static _isFrameRootRequestAndFinished(record) {
+    const isFrameRootRequest = record.url === record.documentURL;
+    const responseReceived = record.responseReceivedTime > 0;
```

Comment:
I really hate trying to remember the difference between data-received/response-received/finished/etc (as well as where things like requestServedFromCache fall into things), but just to make sure I understand, typically `responseReceived` only fires once, we just usually expect a `loadingFinished` to also be fired?

Reply:
Correct that is my understanding as well, data received can fire lots of times, but we always expect one of response received + [loadingFinished/loadingFailed](https://cs.chromium.org/chromium/src/third_party/blink/renderer/devtools/front_end/sdk/NetworkManager.js?type=cs&q=loadingFailed+f:devtools+lang:js&sq=package:chromium&g=0&l=841-843) to signal the end of the request.

Reply:
do we need to consider preloads or won't this affect anything?